From 955dba8428122e14330c7af9f1fb2edf94a23ae7 Mon Sep 17 00:00:00 2001
From: Yu Qiaowei <cerf.yu@rock-chips.com>
Date: Wed, 27 Jan 2021 10:41:32 +0800
Subject: [PATCH] video/rockchip: rga2: Fix the memory leak in rga2 driver.

1. Fix that when fd is used in A+B->C mode, the fd of
   pat is not release.
2. Modify that the pat import from the application is
   only copy to pat or src1.

Signed-off-by: Yu Qiaowei <cerf.yu@rock-chips.com>
Change-Id: I360459ba85fdaad5536a3317b5f415298be87d6b
---
 drivers/video/rockchip/rga2/rga2.h          |  2 ++
 drivers/video/rockchip/rga2/rga2_drv.c      | 24 +++++++++++++++------
 drivers/video/rockchip/rga2/rga2_reg_info.c |  8 +++++--
 3 files changed, 26 insertions(+), 8 deletions(-)

diff --git a/drivers/video/rockchip/rga2/rga2.h b/drivers/video/rockchip/rga2/rga2.h
index ca3fa9c05d18..eb63f6f58de2 100644
--- a/drivers/video/rockchip/rga2/rga2.h
+++ b/drivers/video/rockchip/rga2/rga2.h
@@ -693,10 +693,12 @@ struct rga2_reg {
 	struct sg_table *sg_src0;
 	struct sg_table *sg_src1;
 	struct sg_table *sg_dst;
+	struct sg_table *sg_els;
 
 	struct dma_buf_attachment *attach_src0;
 	struct dma_buf_attachment *attach_src1;
 	struct dma_buf_attachment *attach_dst;
+	struct dma_buf_attachment *attach_els;
 };
 
 struct rga2_service_info {
diff --git a/drivers/video/rockchip/rga2/rga2_drv.c b/drivers/video/rockchip/rga2/rga2_drv.c
index ca2148506c35..8d0ef9787aee 100644
--- a/drivers/video/rockchip/rga2/rga2_drv.c
+++ b/drivers/video/rockchip/rga2/rga2_drv.c
@@ -873,9 +873,11 @@ static struct rga2_reg * rga2_reg_init(rga2_session *session, struct rga2_req *r
 	reg->sg_src0 = req->sg_src0;
 	reg->sg_dst = req->sg_dst;
 	reg->sg_src1 = req->sg_src1;
+	reg->sg_els = req->sg_els;
 	reg->attach_src0 = req->attach_src0;
 	reg->attach_dst = req->attach_dst;
 	reg->attach_src1 = req->attach_src1;
+	reg->attach_els = req->attach_els;
 #endif
 
     mutex_lock(&rga2_service.lock);
@@ -1004,8 +1006,8 @@ static int rga2_put_dma_buf(struct rga2_req *req, struct rga2_reg *reg)
 	if (!req && !reg)
 		return -EINVAL;
 
-	attach = (!reg) ? req->attach_src0 : reg->attach_src0;
-	sgt = (!reg) ? req->sg_src0 : reg->sg_src0;
+	attach = reg ? reg->attach_src0 : req->attach_src0;
+	sgt = reg ? reg->sg_src0 : req->sg_src0;
 	if (attach && sgt)
 		dma_buf_unmap_attachment(attach, sgt, DMA_BIDIRECTIONAL);
 	if (attach) {
@@ -1014,8 +1016,8 @@ static int rga2_put_dma_buf(struct rga2_req *req, struct rga2_reg *reg)
 		dma_buf_put(dma_buf);
 	}
 
-	attach = (!reg) ? req->attach_dst : reg->attach_dst;
-	sgt = (!reg) ? req->sg_dst : reg->sg_dst;
+	attach = reg ? reg->attach_dst : req->attach_dst;
+	sgt = reg ? reg->sg_dst : req->sg_dst;
 	if (attach && sgt)
 		dma_buf_unmap_attachment(attach, sgt, DMA_BIDIRECTIONAL);
 	if (attach) {
@@ -1024,8 +1026,18 @@ static int rga2_put_dma_buf(struct rga2_req *req, struct rga2_reg *reg)
 		dma_buf_put(dma_buf);
 	}
 
-	attach = (!reg) ? req->attach_src1 : reg->attach_src1;
-	sgt = (!reg) ? req->sg_src1 : reg->sg_src1;
+	attach = reg ? reg->attach_src1 : req->attach_src1;
+	sgt = reg ? reg->sg_src1 : req->sg_src1;
+	if (attach && sgt)
+		dma_buf_unmap_attachment(attach, sgt, DMA_BIDIRECTIONAL);
+	if (attach) {
+		dma_buf = attach->dmabuf;
+		dma_buf_detach(dma_buf, attach);
+		dma_buf_put(dma_buf);
+	}
+
+	attach = reg ? reg->attach_els : req->attach_els;
+	sgt = reg ? reg->sg_els : req->sg_els;
 	if (attach && sgt)
 		dma_buf_unmap_attachment(attach, sgt, DMA_BIDIRECTIONAL);
 	if (attach) {
diff --git a/drivers/video/rockchip/rga2/rga2_reg_info.c b/drivers/video/rockchip/rga2/rga2_reg_info.c
index 756e6042f509..d4b7e13b41c4 100644
--- a/drivers/video/rockchip/rga2/rga2_reg_info.c
+++ b/drivers/video/rockchip/rga2/rga2_reg_info.c
@@ -1091,8 +1091,12 @@ void RGA_MSG_2_RGA2_MSG(struct rga_req *req_rga, struct rga2_req *req)
 
     memcpy(&req->src, &req_rga->src, sizeof(req_rga->src));
     memcpy(&req->dst, &req_rga->dst, sizeof(req_rga->dst));
-    memcpy(&req->pat, &req_rga->pat, sizeof(req_rga->pat));
-    memcpy(&req->src1,&req_rga->pat, sizeof(req_rga->pat));
+    /* The application will only import pat or src1. */
+    if (req->render_mode == update_palette_table_mode) {
+        memcpy(&req->pat, &req_rga->pat, sizeof(req_rga->pat));
+    } else {
+        memcpy(&req->src1, &req_rga->pat, sizeof(req_rga->pat));
+    }
 
     format_name_convert(&req->src.format, req_rga->src.format);
     format_name_convert(&req->dst.format, req_rga->dst.format);
-- 
2.35.3

